File: 13. kmeans.py
# -*- coding: utf-8 -*-
"""
Created on Thu Aug 19 08:53:32 2021

@author: Admin
"""

## Kmeans##

# =============================================================================
# libraries
# =============================================================================

import pandas as pd
import matplotlib.pyplot as plt
# import sklearn.preprocessing as sklp
import sklearn.cluster as sklc

# =============================================================================
# read data
# =============================================================================

mall = pd.read_csv("data\Mall_Customers.csv")

x = mall.iloc[:, [3, 4]].values #ingresos y puntuacion

# =============================================================================
# método del codo
# =============================================================================

wcss = []

for i in range(1,11):
    
    kmeans = sklc.KMeans(n_clusters = i, init = "k-means++", 
                         max_iter = 300, n_init = 10, random_state = 0)
    
    kmeans.fit(x)
    
    wcss.append(kmeans.inertia_)

plt.plot(range(1,11), wcss)
plt.title("Método del codo")
plt.xlabel("Número de cluster")
plt.ylabel("WCSS")    
plt.show()
    
# =============================================================================
# Ajustar modelo de cluster
# =============================================================================

kmeans = sklc.KMeans(n_clusters = 5, init = "k-means++", 
                         max_iter = 300, n_init = 10, random_state = 0)

y_cluster = kmeans.fit_predict(x)

# =============================================================================
# Visualización de los cluster
# =============================================================================

plt.scatter(x[y_cluster == 0, 0], x[y_cluster == 0, 1], 
            c = "darkred", s = 100, label = "Cluster 1")
plt.scatter(x[y_cluster == 1, 0], x[y_cluster == 1, 1], 
            c = "darkblue", s = 100, label = "Cluster 2")
plt.scatter(x[y_cluster == 2, 0], x[y_cluster == 2, 1], 
            c = "green", s = 100, label = "Cluster 3")
plt.scatter(x[y_cluster == 3, 0], x[y_cluster == 3, 1], 
            c = "cyan", s = 100, label = "Cluster 4")
plt.scatter(x[y_cluster == 4, 0], x[y_cluster == 4, 1], 
            c = "magenta", s = 100, label = "Cluster 5")
plt.scatter(kmeans.cluster_centers_[:, 0], kmeans.cluster_centers_[:, 1], 
            c = "black", s = 300, label = "Centroides")
plt.title("Cluster de clientres")
plt.xlabel("Ingresos anuales (miles $)")
plt.ylabel("Puntuación de gastos")
plt.legend()
plt.show()

File: 13. kmeans.R
## K-means ##


# libraries ---------------------------------------------------------------

library(tidyverse)
library(cluster)


# read data ---------------------------------------------------------------

mall <- readr::read_csv("data/Mall_Customers.csv") %>% 
  janitor::clean_names()

mall <- mall %>% 
  dplyr::select(annual_income_k, spending_score_1_100)


# método del codo ---------------------------------------------------------

set.seed(123)

codo <- function(i){
  
  x <- mall %>% 
    kmeans(i)
  
  wcss <- sum(x$withinss)
  
}

wcss <- c(1:10) %>% 
  purrr::map_dbl(codo)

ggplot2::ggplot(data = NULL, aes(c(1:10), wcss)) + 
  geom_line(color = "darkblue") +
  geom_point(color = "darkblue") +
  labs(title = "Método del codo",
       x = "Número de k's",
       y = "WCSS") +
  theme_classic()


# ajustar algoritmo de cluster --------------------------------------------

cluster <- mall %>% 
  kmeans(5, iter.max = 300, nstart = 10)


# visualización -----------------------------------------------------------

## Ojo, hace un PCA, o similar

cluster::clusplot(mall, cluster$cluster, 
                  lines = 0, shade = TRUE, 
                  color = TRUE)
